kernel/src/dynamic_process_loading.rs

```diff
@@ -0,0 +1,991 @@
+// Licensed under the Apache License, Version 2.0 or the MIT License.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// Copyright Tock Contributors 2024.
+
+//! Dynamic Process Loader for OTA application loads and updates
+//!
+//! These functions facilitate dynamic application loading and process
+//! creation during runtime without requiring the user to restart the
+//! device.
+
+use core::cell::Cell;
+use core::cmp;
+
+use crate::capabilities::ProcessManagementCapability;
+use crate::config;
+use crate::create_capability;
+use crate::debug;
+use crate::hil::nonvolatile_storage::{NonvolatileStorage, NonvolatileStorageClient};
+use crate::kernel::Kernel;
+use crate::platform::chip::Chip;
+use crate::process::ProcessId;
+use crate::process::{self, Process, ShortId};
+use crate::process_binary::{ProcessBinary, ProcessBinaryError};
+use crate::process_loading::ProcessLoadError;
+use crate::process_policies::ProcessFaultPolicy;
+use crate::process_standard::ProcessStandard;
+use crate::utilities::cells::{MapCell, OptionalCell, TakeCell};
+use crate::ErrorCode;
+
+const TBF_HEADER_LENGTH: usize = 16;
+pub const BUF_LEN: usize = 512;
+const MAX_PROCS: usize = 10; //need this to store the start addresses of processes to write padding
+
+#[derive(Clone, Copy, PartialEq)]
+pub enum NonvolatileCommand {
+    UserspaceRead,
+    UserspaceWrite,
+}
+
+#[derive(Clone, Copy)]
+pub enum State {
+    Setup,
+    AppWrite,
+    Load,
+    PaddingWrite,
+}
+
+/// This interface supports loading processes at runtime.
+pub trait DynamicProcessLoading {
+    /// Call to request loading a new process.
+    ///
+    /// This informs the kernel we want to load a process and the size of the entire process binary.
+    /// The kernel will try to find a suitable location in flash to store said process.
+    ///
+    /// Return value:
+    /// - `Ok((start_address, length))`: If there is a place to load the
+    ///   process, the function will return `Ok()` with the address to start at
+    ///   and the size of the region to store the process.
+    /// - `Err(ErrorCode)`: If there is nowhere to store the process a suitable
+    ///   `ErrorCode` will be returned.
+    fn setup(&self, app_length: usize) -> Result<usize, ErrorCode>;
+
+    /// Instruct the kernel to write data to the flash
+    ///
+    ///This is used to write both userland apps and padding apps
+    fn write_app_data(
+        &self,
+        buffer: &'static mut [u8],
+        offset: usize,
+        app_size: usize,
```

Comment:
Can you replace `buffer` and `app_size` with a `SubSliceMut`?

Reply:
Do we want the capsule to do the sub slicing?

Reply:
Yes